plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/delete/DeletionVectors.java

```diff
@@ -67,6 +75,56 @@ public static Roaring64NavigableMap readDeletionVectors(TrinoFileSystem fileSyst
         throw new IllegalArgumentException("Unexpected storage type: " + deletionVector.storageType());
     }
 
+    public static DeletionVectorEntry writeDeletionVectors(
+            TrinoFileSystem fileSystem,
+            Location location,
+            Roaring64Bitmap pastDeletionVectors,
+            LongBitmapDataProvider rowsDeletedByDelete,
+            LongBitmapDataProvider rowsDeletedByUpdate)
+            throws IOException
+    {
+        UUID uuid = randomUUID();
+        String deletionVectorFilename = "deletion_vector_" + uuid + ".bin";
+        String pathOrInlineDv = encodeUuid(uuid);
+        RoaringBitmap bitmap = toRoaringBitmap(pastDeletionVectors, rowsDeletedByDelete, rowsDeletedByUpdate);
+        int sizeInBytes = MAGIC_NUMBER_BYTE_SIZE + BIT_MAP_COUNT_BYTE_SIZE + BIT_MAP_KEY_BYTE_SIZE + bitmap.serializedSizeInBytes();
+        long cardinality = pastDeletionVectors.getIntCardinality() + rowsDeletedByDelete.getLongCardinality() + rowsDeletedByUpdate.getLongCardinality();
+
+        checkArgument(sizeInBytes > 0, "sizeInBytes must be positive: %s", sizeInBytes);
+        checkArgument(cardinality > 0, "cardinality must be positive: %s", cardinality);
+
+        OptionalInt offset;
+        byte[] data = serializeAsByteArray(bitmap, sizeInBytes);
+        try (DataOutputStream output = new DataOutputStream(fileSystem.newOutputFile(location.appendPath(deletionVectorFilename)).create())) {
+            output.writeByte(FORMAT_VERSION_V1);
+            offset = OptionalInt.of(output.size());
+            output.writeInt(sizeInBytes);
+            output.write(data);
+            output.writeInt(calculateChecksum(data));
+        }
+
+        return new DeletionVectorEntry(UUID_MARKER, pathOrInlineDv, offset, sizeInBytes, cardinality);
+    }
+
+    private static byte[] serializeAsByteArray(RoaringBitmap bitmap, int sizeInBytes)
+    {
+        ByteBuffer buffer = ByteBuffer.allocate(sizeInBytes).order(LITTLE_ENDIAN);
+        buffer.putInt(PORTABLE_ROARING_BITMAP_MAGIC_NUMBER);
+        buffer.putLong(1); // Always write single RoaringBitmap
+        buffer.putInt(0); // Bitmap index
+        bitmap.serialize(buffer);
+        return buffer.array();
+    }
+
+    private static RoaringBitmap toRoaringBitmap(Roaring64Bitmap pastDeletionVectors, LongBitmapDataProvider rowsDeletedByDelete, LongBitmapDataProvider rowsDeletedByUpdate)
+    {
+        RoaringBitmap roaringBitmap = new RoaringBitmap();
+        pastDeletionVectors.forEach(position -> roaringBitmap.add(toIntExact(position)));
+        rowsDeletedByDelete.stream().forEach(position -> roaringBitmap.add(toIntExact(position)));
+        rowsDeletedByUpdate.stream().forEach(position -> roaringBitmap.add(toIntExact(position)));
```

Comment:
OR these bitmaps using RoaringBitmap API
You can probably use this method in `isNoRowsRetained` as well

Reply:
This return type is `RoaringBitmap` because Delta Lake doesn't support reading `Roaring64Bitmap` as far as I confirmed the code, and the arguments are  `Roaring64Bitmap` types.

I think we can't use OR in this situation?

Reply:
The description at https://github.com/delta-io/delta/blob/master/PROTOCOL.md#Deletion-Vector-Format suggests that the serialization of RoaringBitmap portable between 32 bit and 64 bit versions.
The spec is also clearly saying that "Deletion Vectors use 64-bit integers".
The `toIntExact` checks here look problematic, i think these could fail when we're dealing with files containing more than Integer.MAX_VALUE rows.

Reply:
It's just a documentation, not code. https://github.com/delta-io/delta/blob/master/kernel/kernel-api/src/main/java/io/delta/kernel/internal/deletionvectors/RoaringBitmapArray.java and https://github.com/delta-io/delta/blob/master/spark/src/main/scala/org/apache/spark/sql/delta/deletionvectors/RoaringBitmapArray.scala are the relevant classes. 

I confirmed Spark throws an exception if we use Roaring64Bitmap.

Reply:
What happens when there is a file with more than Integer.MAX_VALUE rows and a position bigger than Integer.MAX_VALUE rows is deleted ?
It looks like we will fail the `toIntExact` check, what will Spark do in that case ?
The code in delta-io/Spark looks like a custom implementation of Roaring64Bitmap which still supports larger than Integer.MAX_VALUE values without explicitly using Roaring64Bitmap. It's maintaining an array of RoaringBitmap internally and de-/serializing that.
Their max representable value https://github.com/delta-io/delta/blob/455dbacaf456158644bcbe8b0b13bd4eae65c974/spark/src/main/scala/org/apache/spark/sql/delta/deletionvectors/RoaringBitmapArray.scala#L462 looks bigger than Integer.MAX_VALUE.
Maybe we should just copy that class and it every where consistently.

Reply:
Added `RoaringBitmapArray` class. I didn't copy the entire class because the required logic are limited.

Reply:
The usage of RoaringBitmapArray class now is good, it avoids the toIntExact problem.
It would be even better if we could switch to that class entirely instead of Roaring64Bitmap as RoaringBitmapArray is more optimized for our case where values above `Integer.MAX_VALUE` are possible but rare.
It would also avoid the need to convert from Roaring64Bitmap to RoaringBitmapArray.